File: 0020-valid-parentheses/0020-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
function isValid(s) {
  const openingBrackets = ['(', '[', '{'];
  const closingBrackets = [')', ']', '}'];
  const stack = [];

  for (let i = 0; i < s.length; i++) {
    if (openingBrackets.includes(s[i])) {
      stack.push(s[i]);
    } else if (closingBrackets.includes(s[i])) {
      const expectedOpeningBracket = openingBrackets[closingBrackets.indexOf(s[i])];
      if (stack.pop() !== expectedOpeningBracket) {
        return false;
      }
    }
  }

  return stack.length === 0;
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
         opening_brackets = ['(', '[', '{']
         closing_brackets = [')', ']', '}']
         stack = []

         for char in s:
            if char in opening_brackets:
                stack.append(char)
            elif char in closing_brackets:
                expected_opening = opening_brackets[closing_brackets.index(char)]
                if not stack or stack.pop() != expected_opening:
                    return False

         return len(stack) == 0